File: android/src/main/java/com/maomengte/alibc4/Alibc4Plugin.java
package com.maomengte.alibc4;

import android.app.Activity;
import android.app.Application;
import android.content.Context;
import android.text.TextUtils;
import android.webkit.WebChromeClient;
import android.webkit.WebViewClient;

import androidx.annotation.NonNull;

import com.ali.auth.third.core.MemberSDK;
import com.ali.auth.third.core.config.AuthOption;
import com.ali.auth.third.core.context.KernelContext;
import com.ali.auth.third.core.model.Session;
import com.ali.auth.third.ui.LoginActivity;
import com.alibaba.baichuan.android.trade.AlibcTrade;
import com.alibaba.baichuan.android.trade.AlibcTradeSDK;
import com.alibaba.baichuan.android.trade.callback.AlibcTradeInitCallback;
import com.alibaba.baichuan.android.trade.model.AlibcShowParams;
import com.alibaba.baichuan.android.trade.model.OpenType;
import com.alibaba.baichuan.android.trade.page.AlibcAddCartPage;
import com.alibaba.baichuan.android.trade.page.AlibcBasePage;
import com.alibaba.baichuan.android.trade.page.AlibcDetailPage;
import com.alibaba.baichuan.android.trade.page.AlibcMyCartsPage;
import com.alibaba.baichuan.android.trade.page.AlibcShopPage;
import com.alibaba.baichuan.trade.biz.AlibcConstants;
import com.alibaba.baichuan.trade.biz.AlibcTradeCallback;
import com.alibaba.baichuan.trade.biz.applink.adapter.AlibcFailModeType;
import com.alibaba.baichuan.trade.biz.auth.AlibcAuth;
import com.alibaba.baichuan.trade.biz.context.AlibcTradeResult;
import com.alibaba.baichuan.trade.biz.core.taoke.AlibcTaokeParams;
import com.alibaba.baichuan.trade.biz.login.AlibcLogin;
import com.alibaba.baichuan.trade.biz.login.AlibcLoginCallback;
import com.alibaba.baichuan.trade.common.AlibcTradeCommon;
import com.ut.device.UTDevice;

import java.util.HashMap;
import java.util.Map;
import java.util.Objects;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.embedding.engine.plugins.activity.ActivityAware;
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;

/**
 * Alibc4Plugin
 */
public class Alibc4Plugin implements FlutterPlugin, MethodCallHandler, ActivityAware {
    /// The MethodChannel that will the communication between Flutter and native Android
    ///
    /// This local reference serves to register the plugin with the Flutter Engine and unregister it
    /// when the Flutter Engine is detached from the Activity
    private MethodChannel channel;
    Context context;
    Activity activity;
    AlibcShowParams showParams;
    AlibcTaokeParams taokeParams;
    Map<String, String> trackParams = new HashMap<>(); //自定义参数

    @Override
    public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
        context = flutterPluginBinding.getApplicationContext();
        channel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "alibc4");
        channel.setMethodCallHandler(this);
    }


    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
        if (call.method.equals("getPlatformVersion")) {
            result.success("Android " + android.os.Build.VERSION.RELEASE);
        } else if (call.method.equals("init")) {
            AlibcTradeSDK.asyncInit((Application) context, new AlibcTradeInitCallback() {
                @Override
                public void onSuccess() {
                    result.success(getResult(0));
                }

                @Override
                public void onFailure(int code, String msg) {
                    result.success(getResult(-1, code + " " + msg));
                }
            });
        } else if (call.method.equals("login")) {
            AlibcLogin.getInstance().turnOnDebug();
            AlibcLogin.getInstance().showLogin(new AlibcLoginCallback() {
                @Override
                public void onSuccess(int loginResult, String s, String openId) {
                    LogUtils.e("login>>" + openId);
                    result.success(getResult(0, openId));
                }

                @Override
                public void onFailure(int code, String msg) {
                    // code：错误码  msg： 错误信息
                    result.success(getResult(-1, code + " " + msg));
                }
            });
        } else if (call.method.equals("logout")) {
            AlibcLogin.getInstance().logout(new AlibcLoginCallback() {
                @Override
                public void onSuccess(int loginResult, String s, String openId) {
                    LogUtils.e("logout>>" + openId);
                    result.success(getResult(0));
                }

                @Override
                public void onFailure(int code, String msg) {
                    // code：错误码  msg： 错误信息
                    result.success(getResult(-1, code + " " + msg));
                }
            });
        } else if (call.method.equals("getUserInfo")) {
            Session session = AlibcLogin.getInstance().getSession();
            if (session != null) {
                Map<String, Object> map = new HashMap<>();
                map.put("userid", session.userid);
                map.put("avatarUrl", session.avatarUrl);
                map.put("openId", session.openId);
                map.put("openSid", session.openSid);
                map.put("topAccessToken", session.topAccessToken);
                map.put("topAuthCode", session.topAuthCode);
                map.put("topExpireTime", session.topExpireTime);
                map.put("ssoToken", session.ssoToken);
                map.put("havanaSsoToken", session.havanaSsoToken);
                result.success(getResult(0, map));
            } else {
                result.success(getResult(-1));
            }
        } else if (call.method.equals("setChannel")) {
            Map<String, String> map = (Map<String, String>) call.arguments;
            AlibcTradeSDK.setChannel(map.get("typeName"), map.get("channelName"));
            result.success(getResult(0));
        } else if (call.method.equals("setISVVersion")) {
            AlibcTradeSDK.setISVVersion((String) call.arguments);
        } else if (call.method.equals("openByBizCode")) {
            Map<String, String> map = (Map<String, String>) call.arguments;
            initParams(map);
            AlibcBasePage page = null;
            String code = "detail";
            if (map.get("itemId") != null) {//打开详情页
                page = new AlibcDetailPage(map.get("itemId"));
                code = "detail";
            }
            if (map.get("shopId") != null) {//打开店铺首页
                page = new AlibcShopPage(map.get("shopId"));
                code = "shop";
            }
            AlibcTrade.openByBizCode(activity, page, null, new WebViewClient(), new WebChromeClient(), code, showParams, taokeParams, trackParams, new AlibcTradeCallback() {
                @Override
                public void onTradeSuccess(AlibcTradeResult tradeResult) {
                    result.success(getResult(0));
                }

                @Override
                public void onFailure(int code, String msg) {
                    result.success(getResult(-1, code + " " + msg));
                }
            });
        } else if (call.method.equals("openCart")) {
            Map<String, String> map = (Map<String, String>) call.arguments;
            initParams(map);
            AlibcBasePage page = new AlibcMyCartsPage();
            AlibcTrade.openByBizCode(activity, page, null, new WebViewClient(), new WebChromeClient(), "cart", showParams, taokeParams, trackParams, new AlibcTradeCallback() {
                @Override
                public void onTradeSuccess(AlibcTradeResult tradeResult) {
                    result.success(getResult(0));
                }

                @Override
                public void onFailure(int code, String msg) {
                    result.success(getResult(-1, code + " " + msg));
                }
            });
        } else if (call.method.equals("openByUrl")) {
            Map<String, String> map = (Map<String, String>) call.arguments;
            initParams(map);
            if (map.get("url") != null) {
                AlibcTrade.openByUrl(activity, "", map.get("url"), null, new WebViewClient(), new WebChromeClient(), showParams, taokeParams, trackParams, new AlibcTradeCallback() {
                    @Override
                    public void onTradeSuccess(AlibcTradeResult tradeResult) {
                        result.success(getResult(0));
                    }

                    @Override
                    public void onFailure(int code, String msg) {
                        result.success(getResult(-1, code + " " + msg));
                    }
                });
            } else {
                result.success(getResult(-1, "url must not be null"));
            }
        } else if (call.method.equals("checkSession")) {
            AlibcLogin alibcLogin = AlibcLogin.getInstance();
            if (alibcLogin.isLogin()) {
                result.success(getResult(0, alibcLogin.getSession().openId));
            } else {
                result.success(getResult(0));
            }
        } else if (call.method.equals("getUtdid")) {
            result.success(getResult(0, UTDevice.getUtdid(context)));
        } else {
            result.notImplemented();
        }
    }

    void initParams(Map map) {
        LogUtils.e("initParams>>" + map.toString());
        showParams = new AlibcShowParams();
        //OpenType（页面打开方式）： 枚举值（Auto和Native），Native表示唤端，Auto表示不做设置
        if (map.get("openType") != null && map.get("openType").equals("auto")) {
            showParams.setOpenType(OpenType.Auto);
        } else {
            showParams.setOpenType(OpenType.Native);
        }
        //clientType表示唤端类型：taobao---唤起淘宝客户端；tmall---唤起天猫客户端
        if (map.get("clientType") != null && map.get("clientType").equals("tmall")) {
            showParams.setClientType("tmall");
        } else {
            showParams.setClientType("taobao");
        }
        showParams.setOpenType(OpenType.Native);
        showParams.setClientType("taobao");
        if (map.get("title") != null) {
            showParams.setTitle(map.get("title").toString());
        }
        if (map.get("degradeUrl") != null) {
            showParams.setDegradeUrl(map.get("degradeUrl").toString());
        }
        //BACK_URL（小把手）：唤端返回的scheme,如果不传默认将不展示小把手；如果想展示小把手，可以自己传入自定义的scheme
        if (map.get("backUrl") != null) {
            showParams.setBackUrl(map.get("backUrl").toString());
        }
        int failType = 0;
        if (map.get("failType") != null) {
            failType = (int) map.get("failType");
        }
        //唤端失败模式
        switch (failType) {
            case 0:
            default://默认
                showParams.setNativeOpenFailedMode(AlibcFailModeType.AlibcNativeFailModeJumpH5);
                break;
            case 1://跳转浏览器
                showParams.setNativeOpenFailedMode(AlibcFailModeType.AlibcNativeFailModeJumpBROWER);
                break;
            case 2://跳转下载页
                showParams.setNativeOpenFailedMode(AlibcFailModeType.AlibcNativeFailModeJumpDOWNLOAD);
                break;
        }
        // taokeParams（淘客）参数配置：配置aid或pid的方式分佣
        //（注：1、如果走adzoneId的方式分佣打点，需要在extraParams中显式传入taokeAppkey，否则打点失败；
        // 2、如果是打开店铺页面(shop)，需要在extraParams中显式传入sellerId，否则同步打点转链失败）
        taokeParams = new AlibcTaokeParams("", "", "");
        if (map.get("pid") != null) {
            taokeParams.setPid(map.get("pid").toString());
        }
        if (map.get("aid") != null) {
            taokeParams.setAdzoneid(map.get("aid").toString());
        }
        if (map.get("subPid") != null) {
            taokeParams.setSubPid(map.get("subPid").toString());
        }
        if (map.get("unionId") != null) {
            taokeParams.setUnionId(map.get("unionId").toString());
        }
        taokeParams.extraParams = trackParams;
        if (map.get("sellerId") != null) {
            taokeParams.extraParams.put("sellerId", map.get("sellerId").toString());
        }
        if (map.get("taokeAppkey") != null) {
            taokeParams.extraParams.put("taokeAppkey", map.get("taokeAppkey").toString());
        }
    }

    @Override
    public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
        channel.setMethodCallHandler(null);
    }

    Map<String, Object> getResult(int code) {
        Map<String, Object> map = new HashMap<>();
        map.put("code", code);
        return map;
    }

    Map<String, Object> getResult(int code, Object msg) {
        Map<String, Object> map = new HashMap<>();
        map.put("code", code);
        map.put("msg", msg);
        return map;
    }

    @Override
    public void onAttachedToActivity(@NonNull ActivityPluginBinding binding) {
        activity = binding.getActivity();
    }

    @Override
    public void onDetachedFromActivityForConfigChanges() {

    }

    @Override
    public void onReattachedToActivityForConfigChanges(@NonNull ActivityPluginBinding binding) {

    }

    @Override
    public void onDetachedFromActivity() {

    }
}

File: ios/Classes/Alibc4Plugin.m
#import "Alibc4Plugin.h"
#import "AlibcTradeSDK/AlibcTradeSDK.h"
#import <AlibcTradeSDK/AlibcTradeService.h>
#import <AlibcTradeSDK/AlibcTradePageFactory.h>
#import <AlibabaAuthEntrance/ALBBSDK.h>
#import <AlibabaAuthEntrance/ALBBCompatibleSession.h>
#import <UTDID/UTDevice.h>
#import "webview/ALiTradeWebViewController.h"

@interface Alibc4Plugin ()
@property (nonatomic, strong) AlibcTradeShowParams *showParams;
@property (nonatomic, strong) AlibcTradeTaokeParams *taokeParams;
@end

@implementation Alibc4Plugin

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
  FlutterMethodChannel* channel = [FlutterMethodChannel
      methodChannelWithName:@"alibc4"
            binaryMessenger:[registrar messenger]];
  Alibc4Plugin* instance = [[Alibc4Plugin alloc] init];
  [registrar addMethodCallDelegate:instance channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
    
  if ([@"getPlatformVersion" isEqualToString:call.method]) {
      result(@"getPlatformVersion");
  } else  if ([@"init" isEqualToString:call.method]) {
      // 百川平台基础SDK初始化，加载并初始化各个业务能力插件
      [[AlibcTradeSDK sharedInstance] setDebugLogOpen:YES];
      [[AlibcTradeSDK sharedInstance] asyncInitWithSuccess:^{
          NSLog(@"百川SDK初始化成功");
          result(@{ @"code":@0});
      } failure:^(NSError *error) {
          NSLog(@"百川SDK初始化失败");
          result(@{ @"code":@-1,@"msg":error.description});
      }];
  } else  if ([@"login" isEqualToString:call.method]) {
      if(![[ALBBCompatibleSession sharedInstance] isLogin]) {
          [[ALBBSDK sharedInstance] setH5Only:NO];
          [[ALBBSDK sharedInstance] auth:[UIApplication sharedApplication].keyWindow.rootViewController successCallback:^{
              ALBBUser *user= [[ALBBCompatibleSession sharedInstance] getUser];
              result(@{ @"code":@0,@"msg":user.openId});
          } failureCallback:^(NSError *error) {
              NSString *tip=[NSString stringWithFormat:@"登录失败:%@",error.description];
              result(@{ @"code":@-1,@"msg":tip});
          }];
      } else {
          ALBBUser *user= [[ALBBCompatibleSession sharedInstance] getUser];
          result(@{ @"code":@0,@"msg":user.openId});
      }
  } else  if ([@"logout" isEqualToString:call.method]) {
      if([[ALBBCompatibleSession sharedInstance] isLogin]) {
          [[ALBBSDK sharedInstance] logout];
      }
      result(@{ @"code":@0});
  } else  if ([@"getUserInfo" isEqualToString:call.method]) {
      ALBBUser *user= [[ALBBCompatibleSession sharedInstance] getUser];
      NSDictionary *userInfo=@{
          @"nick":user.nick,
          @"avatarUrl":user.avatarUrl,
          @"openId":user.openId,
          @"openSid":user.openSid,
          @"topAccessToken":user.topAccessToken,
          @"topAuthCode":user.topAuthCode
      };
      result(@{ @"code":@0,@"msg":userInfo});
  }else  if ([@"setChannel" isEqualToString:call.method]) {
      NSDictionary *params=  call.arguments;
      [[AlibcTradeSDK sharedInstance] setChannel:params[@"typeName"] name:params[@"channelName"]];
      result(@{ @"code":@0});
  }else  if ([@"checkSession" isEqualToString:call.method]) {
      if([[ALBBCompatibleSession sharedInstance] isLogin]){
          ALBBUser *user= [[ALBBCompatibleSession sharedInstance] getUser];
          result(@{ @"code":@0,@"msg":user.openId});
      }else{
          result(@{ @"code":@0});
      }
  }else  if ([@"oauth" isEqualToString:call.method]) {

  }else  if ([@"getUtdid" isEqualToString:call.method]) {
      result(@{ @"code":@0,@"msg": [UTDevice utdid]});
  }else  if ([@"setISVVersion" isEqualToString:call.method]) {
      NSString *params=  call.arguments;
      [[AlibcTradeSDK sharedInstance] setIsvVersion:params];
      result(@{ @"code":@0});
  }else  if ([@"openByBizCode" isEqualToString:call.method]) {
      NSDictionary *params=  call.arguments;
      [self initParams:params];
      id<AlibcTradePage> page;
      NSString *code=@"detail";
      if(params[@"shopId"]){
          page = [AlibcTradePageFactory shopPage:params[@"shopId"]];
          code=@"shop";
      }
      if(params[@"itemId"]){
          page = [AlibcTradePageFactory itemDetailPage:params[@"itemId"]];
          code=@"detail";
      }
      //媒体可以传入自定义webview或者不传 百川SDK默认会创建webview
      ALiTradeWebViewController* view = [[ALiTradeWebViewController alloc] init];
      [[AlibcTradeSDK sharedInstance].tradeService
                             openByBizCode:code
                             page:page
                             webView:view.webView
                             parentController:view
                             showParams:_showParams
                             taoKeParams:_taokeParams
                             trackParam:@{}
                             tradeProcessSuccessCallback:^(AlibcTradeResult *alibcTradeResult){
                                   result(@{ @"code":@0});
                            }
                          tradeProcessFailedCallback:^(NSError *error){
                                   result(@{ @"code":@-1,@"msg":error.description});
                          }];
  }else  if ([@"openByUrl" isEqualToString:call.method]) {
      NSDictionary *params=  call.arguments;
      [self initParams:params];
      //媒体可以传入自定义webview或者不传 百川SDK默认会创建webview
      ALiTradeWebViewController* view = [[ALiTradeWebViewController alloc] init];
      [[AlibcTradeSDK sharedInstance].tradeService
                             openByUrl:params[@"url"]
                             identity:@"trade"
                             webView:view.webView
                             parentController:view
                             showParams:_showParams
                             taoKeParams:_taokeParams
                             trackParam:@{}
                             tradeProcessSuccessCallback:^(AlibcTradeResult *alibcTradeResult){
                                   result(@{ @"code":@0});
                            }
                          tradeProcessFailedCallback:^(NSError *error){
                                   result(@{ @"code":@-1,@"msg":error.description});
                          }];
      result(@{ @"code":@0});
  }else  if ([@"openCart" isEqualToString:call.method]) {
      NSDictionary *params=  call.arguments;
      [self initParams:params];
      id<AlibcTradePage> page = [AlibcTradePageFactory myCartsPage];
      NSString *code=@"cart";
      //媒体可以传入自定义webview或者不传 百川SDK默认会创建webview
      ALiTradeWebViewController* view = [[ALiTradeWebViewController alloc] init];
      [[AlibcTradeSDK sharedInstance].tradeService
                             openByBizCode:code
                             page:page
                             webView:view.webView
                             parentController:view
                             showParams:_showParams
                             taoKeParams:_taokeParams
                             trackParam:@{}
                             tradeProcessSuccessCallback:^(AlibcTradeResult *alibcTradeResult){
                                   result(@{ @"code":@0});
                            }
                          tradeProcessFailedCallback:^(NSError *error){
                                   result(@{ @"code":@-1,@"msg":error.description});
                          }];
  }else {
    result(FlutterMethodNotImplemented);
  }
}

-(void)initParams:(NSDictionary*)params{
    NSLog(@"initParams>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>:%@",params);
    _showParams =[[AlibcTradeShowParams alloc] init];
    if(params[@"openType"]&&[params[@"openType"] isEqualToString:@"auto"]){
        _showParams.openType=AlibcOpenTypeAuto;
    }else{
        _showParams.openType=AlibcOpenTypeNative;
    }
    if(params[@"clientType"]&&[params[@"clientType"] isEqualToString:@"tmall"]){
        _showParams.linkKey=@"tmall";
    }else{
        _showParams.linkKey=@"taobao";
    }
    if(params[@"backUrl"]){
        _showParams.backUrl=params[@"backUrl"];
    }
    if(params[@"failType"]){
        _showParams.isNeedCustomNativeFailMode=YES;
        int failType=(int) (long)params[@"failType"];
        switch (failType) {
            case 0:
            default://不做处理
                _showParams.nativeFailMode=AlibcNativeFailModeJumpH5;
                break;
            case 1://跳转浏览器

                break;
            case 2://跳转下载页
                _showParams.nativeFailMode=AlibcNativeFailModeJumpDownloadPage;
                break;
            case 3://应用内webview打开
                _showParams.nativeFailMode=AlibcNativeFailModeJumpH5;
                if(params[@"degradeUrl"]){
                    _showParams.degradeUrl=params[@"degradeUrl"];
                }
                break;
        }
    }else{
        _showParams.nativeFailMode=AlibcNativeFailModeJumpH5;
    }
    
    _taokeParams=[[AlibcTradeTaokeParams alloc] init];
    if(params[@"pid"]){
        _taokeParams.pid=params[@"pid"];
    }
    if(params[@"aid"]){
        _taokeParams.adzoneId=params[@"aid"];
    }
    if(params[@"subPid"]){
        _taokeParams.subPid=params[@"subPid"];
    }
    if(params[@"unionId"]){
        _taokeParams.unionId=params[@"unionId"];
    }
    if(params[@"aid"]){
        _taokeParams.adzoneId=params[@"aid"];
    }
    
    NSMutableDictionary *ext=[NSMutableDictionary dictionaryWithDictionary:@{}];
    if(params[@"sellerId"]){
        ext[@"sellerId"]=params[@"sellerId"];
    }
    if(params[@"taokeAppkey"]){
        ext[@"taokeAppkey"]=params[@"taokeAppkey"];
    }
    _taokeParams.extParams=ext;

}

@end
